### DISCUSSION TITLE
UITableViewCell 안에 있는 UITextField의 텍스트 변화 이벤트를 index와 함께 ViewModel까지 전달하는 방법

### DISCUSSION THREAD

**Discussion by JuTaK97** [wafflestudio/seminar-2023#174] [Q&A]
ViewController에는 tableView와 viewModel이 있습니다.

tableView 안에는 여러 tableViewCell 들이 있을 거고, 각 tableViewCell 안에는 UITextField가 있습니다.

각 UITextField의 텍스트가 바뀌면, ViewModel이 갖고 있는 data list의 특정 index의 어떤 값이 바뀌어야 합니다.

### 생각의 흐름 1
1. tableViewCell 이 UITextFieldDelegate 역할을 하도록 한다.
2. tableViewCell에서, `func textField(_ textField: UITextField, shouldChangeCharactersIn range: NSRange, replacementString string: String) -> Bool` 를 이용해 바뀐 텍스트(textField.text) 를 받아온다.
3. 이제 tableViewCell 에서 viewModel에 접근해서 데이터를 바꿔야 한다.
    - tableViewCellDelegate 같은 건 없는데...
    -  ViewController과 tableViewCell 간에 delegate를 직접 하나 만들어서, tableViewCell이 ViewController에게 자기가 몇 번째 인덱스인지 알려주면 ViewController에서 ViewModel의 데이터를 바꿀 수 있다.
    - 그런데 tableViewCell가 자신이 tableView 내에서 몇 번째 인덱스인지는 어떻게 알지?

### 생각의 흐름 2
1. ViewController가 각 tableViewCell 내의 UITextField의 UITextFieldDelegate 역할을 하도록 한다.
2. 텍스트 변화는 받았지마 역시 tableViewCell가 자신이 tableView 내에서 몇 번째 인덱스인지 알 방법이 없다.

결국 tableViewCell가 자신이 tableView 내에서 몇 번째 인덱스인지 알아야 한다.

생각나는 가능한 방법 : 
UITableViewDataSource에서 `func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell` 할 때 각 tableViewCell의 tag로 indexPath를 저장한다

하지만 이렇게 해도, 해당 이벤트를 준 UITextField가 memo인지 title인지 알 방법이 없다. (이것도 tag로 저장??)

이게 맞나요? 이렇게 괴상한 방법을 써서까지 부자연스럽게 이벤트 전달을 하고 싶지 않습니다. 어떻게 해야 세련되게 할까요...

----------

```
     func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        // ...
        cell.title.tag = indexPath.row
        cell.title.addTarget(self, action: #selector(textFieldDidChangeTitle(_:)), for: .editingChanged)
    }

    @objc func textFieldDidChangeTitle(_ textField: UITextField) {
        viewModel.todos[textField.tag].title = textField.text ?? ""
    }
```
이렇게 하니까 복잡한 delegate 안 써도 되긴 하네요... 하지만 저 tag라는 이상한 방법을 꼭 써야 하는 걸까요?

**Comment by shp7724** [comment]
> 결국 tableViewCell가 자신이 tableView 내에서 몇 번째 인덱스인지 알아야 한다.

Cell 스스로 어떤 Todo 아이템에 대응하는지 알고 있다면 꼭 인덱스를 알아야 할 필요는 없을 것 같아요.
예를 들어 `TodoCell`이 `TodoCellViewModel`(가칭)에 대한 weak reference를 갖고 있다면, 어딘가에 저장되어 있는 Todo 리스트의 몇 번째 아이템을 수정하고 있는지 알 수 있지 않을까요?

---

`tag`는 그런 용도에 사용하는 건 아닌 것 같아요 ㅎㅎ `tag` 사용은 권장하지 않습니다.

**Reply by JuTaK97** [reply]
Cell이 자기가 Todo 리스트에서 몇 번째 아이템과 관련이 있는지 어떻게 알죠...??
안다 해도 그걸 알면 MVVM에 어긋나는 거 아닌가요?

**Reply by shp7724** [reply]
Todo 아이템에 ID가 있다면 간단하지 않나요?
Cell이 안다기보다는 ViewModel 계층을 통하면 알아낼 수 있다는 뜻이었어요
Cell은 ViewModel만 들고있고, textField에 대한 이벤트를 ViewModel로 전달해주는 역할만 해주어야 합니다.

**Reply by JuTaK97** [reply]
😵
